clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+                << ArgTy;
+  }
+  void handleStructType(const FieldDecl *FD, QualType ArgTy) final {
+    if (SemaRef.getASTContext().getLangOpts().SYCLStdLayoutKernelParams &&
+        !ArgTy->isStandardLayoutType())
+      IsInvalid =
+          Diag.Report(FD->getLocation(), diag::err_sycl_non_std_layout_type)
+          << ArgTy;
+    else {
+      CXXRecordDecl *RD = ArgTy->getAsCXXRecordDecl();
+      if (!RD->hasTrivialCopyConstructor())
+
+        IsInvalid =
+            Diag.Report(FD->getLocation(),
+                        diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 0 << ArgTy;
+      else if (!RD->hasTrivialDestructor())
+        IsInvalid =
+            Diag.Report(FD->getLocation(),
+                        diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 1 << ArgTy;
+    }
+  }
+
+  // We should be able to ahndle this, so we made it part of the visitor, but
+  // this is 'to be implemented'.
+  void handleArrayType(const FieldDecl *FD, QualType ArgTy) final {
+    IsInvalid = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+                << ArgTy;
+  }
+
+  void handleOtherType(const FieldDecl *FD, QualType ArgTy) final {
+    IsInvalid = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+                << ArgTy;
+  }
+};
+
+// A type that handles the accessor recursion and acts as a base for
+// SyclKernelDeclCreator. It doesn't 'own' anything other than the KernelObj
+// pointer and functionality required to add a param.
+class SyclKernelDeclBase : public SyclKernelFieldHandler<SyclKernelDeclBase> {
+protected:
+  FunctionDecl *KernelObj;
+  llvm::SmallVectorImpl<QualType> &ArgTys;
+  llvm::SmallVectorImpl<ParmVarDecl *> &Params;
+
+  void addParam(const FieldDecl *FD, QualType ArgTy) {
+    ASTContext &Ctx = SemaRef.getASTContext();
+    // TODO: should we split this function up?  These ops NEED to happen in
+    // lockstep, so leaning toward leaving htis as just a somewhat long function
+    // :/
+
+    // Create a new ParmVarDecl based on the new info.
+    ParamDesc newParamDesc = makeParamDesc(FD, ArgTy);
+    auto *NewParam = ParmVarDecl::Create(
+        SemaRef.getASTContext(), KernelObj, SourceLocation(), SourceLocation(),
+        std::get<1>(newParamDesc), std::get<0>(newParamDesc),
+        std::get<2>(newParamDesc), SC_None, /*DefArg*/ nullptr);
+
+    NewParam->setScopeInfo(0, Params.size());
+    NewParam->setIsUsed();
+
+    Params.push_back(NewParam);
+
+    // Create the new type.
+    ArgTys.push_back(ArgTy);
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, ArgTys, Info);
+
+    // Set type, note until the destructor of the owner of KernelObj is called
+    // (to set the parameters), we cannot access the parameters from KernelObj
+    // without memory problems.
+    KernelObj->setType(FuncType);
```

Comment:
I think we should move this part into another function, because we re-create and re-set kernel function type for each new argument, this is strange and doesn't seem effective. Here we can leave only collecting of argument types in `ArgTys` array.